### PR TITLE
Properly implement track's num concepts

### LINKED ISSUE #156: [Track List and Concept Map] Concept count mismatch
Python shows 62 concepts on the overview page but only 14 on the concept map. Same mismatch for some other tracks (e.g. JavaScript 24 vs. 20).

---

image

---

image

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Closes https://github.com/exercism/v3-beta/issues/156

### CODE REVIEW

File: app/controllers/tracks/concepts_controller.rb [hunk 1]
```diff
@@ -15,7 +15,6 @@ def index
     @concept_map_data[:exercises_data] =
       UserTrack::GenerateExerciseStatusMapping.(@user_track)
 
-    @num_concepts = @track.concepts.count
```
Comment (ErikSchierboom) [inline app/controllers/tracks/concepts_controller.rb:18]: This has been replaced with a call to `user_track.num_concepts_taught` in the template

File: app/models/user_track/external.rb [hunk 2]
```diff
@@ -162,6 +162,11 @@ def unlocked_concept_ids
       []
     end
 
+    memoize
+    def num_concepts_taught
+      concept_exercises.flat_map(&:taught_concepts).uniq.size
```
Comment (iHiD) [inline app/models/user_track/external.rb]: `TaughtConcept.select(:concept_id).distinct.size`

File: app/models/exercise/taught_concept.rb [hunk 3]
```diff
@@ -1,4 +1,16 @@
...
+  after_save_commit do
+    # We're updating in a single query instead of two queries to avoid race-conditions
+    # and using read_committed to avoid deadlocks
+    ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
+      concepts_count_sql = Arel.sql(Exercise::TaughtConcept.joins(:exercise).
+        where(exercise: { track: exercise.track_id, status: %i[beta active] }).
+        select("COUNT(DISTINCT(track_concept_id))").
+        to_sql)
+      Track.where(id: exercise.track_id).update_all("num_concepts = (#{concepts_count_sql})")
+    end
```
Comment (iHiD) [inline app/models/exercise/taught_concept.rb]: Nice :)

Maybe just for readibility:
```suggestion
    ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
      concepts_count_sql = Arel.sql(
        Exercise::TaughtConcept.joins(:exercise).
          where(exercise: { track: exercise.track_id, status: %i[beta active] }).
          select("COUNT(DISTINCT(track_concept_id))").
          to_sql
        )
      Track.where(id: exercise.track_id).update_all("num_concepts = (#{concepts_count_sql})")
    end
```

File: app/models/user_track.rb [hunk 4]
```diff
@@ -19,6 +19,8 @@ class UserTrack < ApplicationRecord
     primary_key: :user_id,
     inverse_of: :user_track
 
+  delegate :num_concepts, to: :track
```
Comment (ErikSchierboom) [inline app/models/user_track.rb:22]: I've decided to also delegate to `Track` in `UserTrack` to ensure that the logged-in and logged-out values are consistent.

**Pull request merged**